akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,159 @@
+   * The status of the unreachable member must be changed to 'Down', which can be done with
+   * this method.
+   */
+  final case class Down(address: Address) extends ClusterCommand
+
+  def createExtension(system: ActorSystem[_]): Cluster = {
+    require(system.isInstanceOf[ActorSystemAdapter[_]], "only adapted actor systems can be used for cluster features")
+    ???
+  }
+}
+
+@DoNotInherit
+sealed trait Cluster extends Extension {
+
+  import Cluster._
+
+  // Note: not sure about this, but when five accessors are prefixed with `self` that kind of points towards the need
+  // for a scope/namespace for them
+
+  trait ClusterSelf {
+    /** The address of this cluster member. */
+    def address: Address
+
+    /** Data center to which this node belongs to (defaults to "default" if not configured explicitly) */
+    def dataCenter: akka.cluster.ClusterSettings.DataCenter
+
+    /** Scala API: roles that this member has */
+    def roles: Set[String]
+
+    // TODO javadsl/scaladsl rather than a coveritall API?
+    /** Java API: roles that this member has */
+    def getRoles: java.util.Set[String]
+
+    // and maybe, just maybe these two belongs here as well, as they are about the cluster node itself -
+    // the self-view of the state and whether the self-has been shutdown?
+    /** Returns true if this cluster instance has be shutdown. */
+    def isTerminated: Boolean
+
+    // also, the mutable snapshot feels icky but I guess it simplifies things enough to keep?
+    /** Current snapshot state of the cluster. */
+    def state: CurrentClusterState
+  }
+
+  /** Details about this cluster node itself */
+  def self: ClusterSelf
+
+  // I'm thinking settings and failure detectors are really implementation details that has leaked
+  // and should not be accessible through the public api?
+
+  /**
+   * @return an actor that allows for subscribing to messages when the cluster state changes
+   */
+  def subscriptions: ActorRef[ClusterStateSubscription]
+
+  /**
+   * @return an actor that accepts commands to join, leave and down nodes in a cluster
+   */
+  def manager: ActorRef[ClusterCommand]
+
+  // These ones I think should be removed in favour of subscribing to events because that makes how it is executed more clear
+  // and also gives less API surface
+  // problem is then that actor may be dead on remove, but that is a general problem for all actors so shouldn't be suprising?
+  // Or we could keep a subscription-actor alive at all times and handle that there?
+  /*
+  def registerOnMemberUp[T](code: ⇒ T): Unit =
+    registerOnMemberUp(new Runnable { def run() = code })
+
+  def registerOnMemberUp(callback: Runnable): Unit =
+    clusterDaemons ! InternalClusterAction.AddOnMemberUpListener(callback)
+
+  def registerOnMemberRemoved[T](code: ⇒ T): Unit =
+    registerOnMemberRemoved(new Runnable { override def run(): Unit = code })
+
+  // also: it's racy?
```

Comment:
yes, and if we make it actor based it might be more natural to expect that a message (the subscription) will not be delivered if the actor is dead. One could watch if need.